### PR TITLE
Support SSL for gRPC backend

### PR DESCRIPTION
**Pull request by qiwzhang**
This is just make ESP code ready to support SSL for grpc backend.

More changes in following PR:
1) change start_esp.py to add ssl related flags and save them to server_config file.
2) add e2e test.

### CODE REVIEW

File: src/nginx/config.cc [hunk 1]
```diff
@@ -507,6 +508,71 @@ bool ngx_esp_read_server_config_from_string(const std::string &str,
...
+  if (!grpc_ssl.cert_chain_file().empty() &&
+      ngx_esp_read_file(grpc_ssl.cert_chain_file().c_str(), cf->pool,
+                        &ngx_ssl->cert_chain) != NGX_OK) {
+    ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                       "Failed to read cert_chain file: %s",
+                       grpc_ssl.cert_chain_file().c_str());
+    return NGX_ERROR;
+  }
+
+  // If priate_key is not empty, neither should cert_chains
```
Comment (kh-chang) [inline src/nginx/config.cc]: maybe
Either both `private_key` and `cert_chains` exist or neither should

File: src/nginx/grpc.cc [hunk 2]
```diff
@@ -76,6 +76,42 @@ std::pair<Status, std::string> GrpcGetBackendAddress(
       Status(NGX_DECLINED, "No GRPC backend address specified"), std::string());
 }
 
+std::shared_ptr<::grpc::ChannelCredentials> CreateChannelCredentials(
+    ngx_http_request_t *r, ngx_esp_loc_conf_t *espcf) {
+  if (espcf->grpc_backend_ssl == nullptr) {
+    ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
```
Comment (kh-chang) [inline src/nginx/grpc.cc]: Are these all errors? 

All of the logging statements in this function are logged with `ngx_log_error`

Comment (qiwzhang) [reply src/nginx/grpc.cc]: it is only log as debug.  don't be confused with ngx_log_error.  it is NGX_LOG_DEBUG.

File: src/nginx/config.cc (hunk 1 above)
Comment (qiwzhang) [reply src/nginx/config.cc]: Done

File: src/nginx/grpc.cc [hunk 3]
```diff
@@ -76,6 +76,42 @@ std::pair<Status, std::string> GrpcGetBackendAddress(
...
+                  "GrpcGetStub: Use insecure channel credentials");
+    return ::grpc::InsecureChannelCredentials();
+  }
+
+  if (espcf->grpc_backend_ssl->use_google_default) {
+    ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
```
Comment (JLXIA) [inline src/nginx/grpc.cc]: Should this be a Error log?

File: src/nginx/grpc.cc [hunk 4]
```diff
@@ -76,6 +76,42 @@ std::pair<Status, std::string> GrpcGetBackendAddress(
...
+    return ::grpc::GoogleDefaultCredentials();
+  }
+
+  auto ssl = ::grpc::SslCredentialsOptions();
+  ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
+                "GrpcGetStub: Use SSL channel credentials");
+
+  if (espcf->grpc_backend_ssl->root_certs.data != nullptr) {
+    // we have custom roots
+    ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
```
Comment (JLXIA) [inline src/nginx/grpc.cc]: also this and the following, seems like a Info log.

File: src/nginx/t/grpc_interop_upstream_ssl.t [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+$t->run_daemon(\&service_control, $t, $ServiceControlPort, 'servicecontrol.log');
+$t->run_daemon(\&ApiManager::grpc_interop_server, $t, "${GrpcBackendPort}", @ssl_args);
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+is($t->waitforsocket("127.0.0.1:${GrpcBackendPort}"), 1, 'GRPC test server socket ready.');
+$t->run();
+is($t->waitforsocket("127.0.0.1:${Http2NginxPort}"), 1, 'Nginx socket ready.');
+
+################################################################################
+my @test_cases = (
```
Comment (JLXIA) [inline src/nginx/t/grpc_interop_upstream_ssl.t]: should there be any negative tests?

File: src/nginx/grpc.cc (hunk 2 above)
Comment (kh-chang) [reply src/nginx/grpc.cc]: I see -- it seems like `ngx_log_error` is used for non-errors as well. 

FYI: There is also `ngx_log_debug` which is just an alias to `ngx_log_error`
https://www.nginx.com/resources/wiki/extending/api/logging/#ngx-log-debug

File: src/nginx/t/grpc_interop_upstream_ssl.t (hunk 5 above)
Comment (qiwzhang) [reply src/nginx/t/grpc_interop_upstream_ssl.t]: Done.  Try to add two negative test cases, only one of them work, another one discovered a bug in grpc c++ code, have to disable that test.

File: src/nginx/grpc.cc (hunk 3 above)
Comment (qiwzhang) [reply src/nginx/grpc.cc]: Not this is a valid use case, not an error.

File: src/nginx/grpc.cc (hunk 4 above)
Comment (qiwzhang) [reply src/nginx/grpc.cc]: Change it to LOG.

File: src/nginx/t/grpc_interop_upstream_ssl.t [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+# the request will hang if channel has credential problems, such as could not read files.
+
+$t->write_file('server_config.pb.txt', <<"EOF");
+grpc_backend_ssl_credentials {
+  use_ssl: true
+}
+EOF
+#run_all_daemons();
+
+#my $result2 = &ApiManager::run_grpc_interop_test($t, $Http2NginxPort, 'empty_unary', '--api_key', 'api-key');
```
Comment (kh-chang) [inline src/nginx/t/grpc_interop_upstream_ssl.t]: nit: forgot to delete these?

**Comment by qiwzhang** [comment]
@kh-chang PTAL

File: src/nginx/t/grpc_interop_upstream_ssl.t (hunk 6 above)
Comment (qiwzhang) [reply src/nginx/t/grpc_interop_upstream_ssl.t]: Not, I deliberately leave it here so when the grpc bug is fixed, we can just un-comment it out.

**Pull request merged**